src/basic/env-util.c

```diff
@@ -1074,3 +1074,17 @@ int getenv_steal_erase(const char *name, char **ret) {
 
         return 1;
 }
+
+int set_full_environment(char **env) {
+        int r;
+
+        clearenv();
+
+        STRV_FOREACH(e, env) {
+                r = putenv(*e);
+                if (r < 0)
+                        return r;
```

Comment:
putenv is really weird, because it makes the specified memory part of environ. That's almost never what we want...

hence use "setenv()", so that glibc gets the memory from wherever it wants to get it from, and we keep our memory.